pkg/providers/cloudstack/validator.go

```diff
@@ -48,83 +55,122 @@ type ProviderCmkClient interface {
 		return err
 	}
 
-	zones, errZone := v.cmk.ValidateZonesPresent(ctx, datacenterConfig.Spec.Zones)
-	if errZone != nil {
-		return fmt.Errorf("checking zones %v", errZone)
-	}
+	for _, az := range v.availabilityZones {
+		_, err := getHostnameFromUrl(az.ManagementApiEndpoint)
+		if err != nil {
+			return fmt.Errorf("checking management api endpoint: %v", err)
+		}
 
-	for _, zone := range datacenterConfig.Spec.Zones {
-		if len(zone.Network.Id) == 0 && len(zone.Network.Name) == 0 {
-			return fmt.Errorf("zone network is not set or is empty")
+		cmk := v.cmks[az.Name]
+		if cmk.GetManagementApiEndpoint(ctx) != az.ManagementApiEndpoint {
+			return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
+				cmk.GetManagementApiEndpoint(ctx), az.ManagementApiEndpoint)
 		}
-		err := v.cmk.ValidateNetworkPresent(ctx, domainId, zone, zones, datacenterConfig.Spec.Account, len(zones) > 1)
+
+		zones, err := cmk.ValidateZonesPresent(ctx, []anywherev1.CloudStackZone{az.CloudStackAvailabilityZone.Zone})
 		if err != nil {
-			return fmt.Errorf("checking network %v", err)
+			return err
+		}
+		if len(zones) > 0 {
+			az.CloudStackAvailabilityZone.Zone.Id = zones[0].Id
+		}
+		if len(az.CloudStackAvailabilityZone.Zone.Network.Id) == 0 && len(az.CloudStackAvailabilityZone.Zone.Network.Name) == 0 {
+			return fmt.Errorf("zone network is not set or is empty")
+		}
+		if err := cmk.ValidateNetworkPresent(ctx, az.DomainId, az.CloudStackAvailabilityZone.Zone, zones, az.Account, true); err != nil {
+			return err
 		}
 	}
 
 	logger.MarkPass("Datacenter validated")
 	return nil
 }
 
-func (v *Validator) validateDomainAndAccount(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
-	if (datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account == "") ||
-		(datacenterConfig.Spec.Domain == "" && datacenterConfig.Spec.Account != "") {
-		return fmt.Errorf("both domain and account must be specified or none of them must be specified")
+func (v *Validator) generateLocalAvailabilityZones(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
+	if datacenterConfig == nil {
+		return nil
 	}
 
-	if datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account != "" {
-		domain, errDomain := v.cmk.ValidateDomainPresent(ctx, datacenterConfig.Spec.Domain)
-		if errDomain != nil {
-			return fmt.Errorf("checking domain: %v", errDomain)
+	if len(datacenterConfig.Spec.Domain) > 0 {
+		cmk := v.cmks[decoder.CloudStackGlobalAZ]
+		domain, err := cmk.ValidateDomainPresent(ctx, datacenterConfig.Spec.Domain)
+		if err != nil {
+			return err
 		}
-
-		errAccount := v.cmk.ValidateAccountPresent(ctx, datacenterConfig.Spec.Account, domain.Id)
-		if errAccount != nil {
-			return fmt.Errorf("checking account: %v", errAccount)
+		if err := cmk.ValidateAccountPresent(ctx, datacenterConfig.Spec.Account, domain.Id); err != nil {
```

Comment:
we don't always need to check that the account is present. The logic is: when account is set, domain must also be set. But domain can also be set without account. See https://github.com/aws/eks-anywhere/blob/main/pkg/providers/cloudstack/validator.go#L103-L123

Reply:
bump